src/psyclone/psyir/nodes/acc_directives.py

```diff
@@ -902,8 +906,90 @@ def _sig_set_to_string(sig_set):
     return ",".join(sorted(names))
 
 
+class ACCAtomicDirective(ACCRegionDirective):
+    '''
+    OpenACC directive to represent that the memory accesses in the associated
+    assignment must be performed atomically.
+
+    '''
+    def begin_string(self):
+        '''
+        :returns: the opening string statement of this directive.
+        :rtype: str
+
+        '''
+        return "acc atomic"
+
+    def end_string(self):
+        '''
+        :returns: the ending string statement of this directive.
+        :rtype: str
+
+        '''
+        return "acc end atomic"
+
+    @staticmethod
+    def is_valid_atomic_statement(stmt):
+        ''' Check if a given statement is a valid OpenACC atomic expression.
+
+        :param stmt: a node to be validated.
+        :type stmt: :py:class:`psyclone.psyir.nodes.Node`
+
+        :returns: whether a given statement is compliant with the OpenACC
+            atomic expression.
+        :rtype: bool
+
+        '''
+        if not isinstance(stmt, Assignment):
+            return False
+
+        # Not all rules are checked, just that:
+        # - operands are of a scalar intrinsic type
+        if not isinstance(stmt.lhs.datatype, ScalarType):
+            return False
+
+        # - operators are one of:
+        #   +, *, -, /, AND, OR, EQV, NEQV
+        if isinstance(stmt.rhs, BinaryOperation):
```

Comment:
No need to walk, it just needs to check the top-level operation, think of it as a combined special operation that exists with the assignment, but what the other operand is doesn't matter (I tested this just in case and in fact the compiler accepts anything).

However, your example make me realize that we at least need to check that one of the top-levels operands is the same as the lhs, so both:
`a = a + (something_not_atomic)`
and
`a = (something_not_atomic) + a`
is supported, but something like:
`a = b + (a + 3)`
is not (I also checked this, the compilers do not rearrange this associative math operation, and instead fails with:
`Error: !$OMP ATOMIC assignment must be var = var op expr or var = expr op var at (1)`)